proof/refine/ARM/Ipc_R.thy

```diff
@@ -2990,6 +2990,163 @@ lemma do_nbrecv_failed_transfer_corres:
+
+lemma valid_tcb_sched_context_update_empty[elim!]:
+  "valid_tcb tp tcb s \<Longrightarrow> valid_tcb tp (tcb_sched_context_update Map.empty tcb) s"
+  by (auto simp: valid_tcb_def tcb_cap_cases_def)
+
+lemma valid_tcb'_SchedContext_update_empty[elim!]:
+  "valid_tcb' tcb s' \<Longrightarrow> valid_tcb' (tcbSchedContext_update Map.empty tcb) s'"
+  by (auto simp: valid_tcb'_def valid_cap'_def tcb_cte_cases_def)
+
+lemma maybeReturnSc_corres:
+  "corres dc
+   (ntfn_at ntfnPtr and tcb_at thread and valid_tcbs and pspace_aligned
+      and scheduler_act_not thread
+      and pspace_distinct and weak_valid_sched_action
+      and not_queued thread and not_in_release_q thread)
+   (valid_tcbs' and valid_queues and valid_queues' and valid_release_queue and valid_release_queue')
+   (maybe_return_sc ntfnPtr thread)
+   (maybeReturnSc ntfnPtr thread)"
+  unfolding maybe_return_sc_def maybeReturnSc_def
+  apply (clarsimp simp: liftM_def get_sk_obj_ref_def get_tcb_obj_ref_def
+                        set_tcb_obj_ref_thread_set)
+  apply (rule stronger_corres_guard_imp)
+    apply (rule corres_split [OF _ get_ntfn_corres])
+      apply (frule ntfn_relation_par_inj[symmetric], simp)
+      apply (rule corres_split [OF _ threadget_corres[where r="(=)"]])
+         apply (rule corres_when2, simp)
+         apply (rule corres_assert_opt_assume_l)
+         apply (rule corres_split [OF _ threadset_corresT])
+              apply (rule_tac Q'="\<top>" in corres_symb_exec_r')
+                 apply (rule corres_split)
+                    prefer 2
+                    apply (rule update_sc_no_reply_stack_update_ko_at'_corres
+                                [where f'="scTCB_update (\<lambda>_. None)"])
+                       apply ((clarsimp simp: sc_relation_def objBits_def objBitsKO_def)+)[4]
+                   apply (rule corres_split [OF _ gct_corres])
+                     apply (rule corres_when [OF _ rescheduleRequired_corres], simp)
+                    apply (wpsimp wp: hoare_vcg_imp_lift')+
+             apply (clarsimp simp: tcb_relation_def)
+            apply (rule ball_tcb_cap_casesI; simp)
+           apply (clarsimp simp: tcb_cte_cases_def)
+          apply (wpsimp wp: hoare_vcg_imp_lift' thread_set_weak_valid_sched_action2)
+         apply (wpsimp wp: hoare_drop_imp threadSet_valid_queues_no_state
+                           threadSet_valid_queues' threadSet_valid_release_queue
+                           threadSet_valid_tcbs'
+                           threadSet_valid_release_queue')
+        apply (clarsimp simp: tcb_relation_def)
+       apply (wpsimp wp: thread_get_wp threadGet_wp)+
+      apply (frule ntfn_relation_par_inj, simp)
+     apply (wpsimp wp: get_simple_ko_wp getNotification_wp)+
+   apply (rule valid_tcbs_valid_tcbE, simp, simp)
+   apply clarsimp
+   apply (intro conjI)
+     apply (clarsimp simp: valid_tcb_def valid_bound_obj_def split: option.splits)
+    apply clarsimp
+   apply clarsimp
```

Comment:
Can the `intro conjI` and the clarsimps be removed and replaced with just the one clarsimp?

Reply:
Weirdly, no. A number of other variation seem not to work.